_modules/ext_module_qvm.py

```diff
@@ -941,8 +941,203 @@ def prefs(vmname, *varargs, **kwargs):
     return qvm.status()
 
 
+def devices(vmname, *varargs, **kwargs):
+    """
+    Manage a virtual machine domain devices::
+
+    Valid actions:
+
+    .. code-block:: yaml
+
+        # Required Positional
+        - name:                 <vmname>
+        - action:               [list|attach|detach]
+        - devices:             [string,]
+
+    Example:
+
+    .. code-block:: yaml
+
+        # List
+        test-vm-1:  # (test-vm-1 is the VM name)
+          qvm.devices:
+            - list: []
+
+        test-vm-2:
+          qvm.devices: []
+
+        some-label-that-is-not-the-vm-name:
+          qvm.devices:
+            - name: test-vm-3
+
+        # Attach, detach
+        test-vm-4:
+          qvm.features:
+            - attach:
+              - pci:dom0:01_00.0: []
+              - pci:dom0:02_00.0:
+                - pci_strictreset: false
+              - bridge:sys-net-interfaces:br0:
+                - ip: 192.168.0.1
+                - netmask: 255.255.255.0
+            - detach:
+              - pci:dom0:28_00.3: []
+              - bridge:sys-net-interfaces:br1: []
+    """
+
+    # CLI 'qubesctl qvm.devices <vm_name> (attach|detach) device [device...]'
+    if varargs and varargs[0] in ['attach', 'detach']:
+        devices = []
+        for device in varargs[1:]:
+            device_type, backend, dev_id = device.split(':')
+            devices.append({'device_type': device_type, 'backend': backend, 'dev_id': dev_id})
+        if devices:
+            kwargs[varargs[0]] = devices
+
+    # Convert kwargs to string for argparse
+    for action in ['attach', 'detach']:
+        if action in kwargs:
+            kwargs[action] = json.dumps(kwargs[action])
+
+    # Set default status-mode to show all status entries
+    kwargs.setdefault('status-mode', 'all')
+
+    qvm = _QVMBase('qvm.devices', **kwargs)
+    qvm.parser.add_argument(
+        'vmname',
+        action=_VMAction,
+        help='Virtual machine name'
+    )
+    qvm.parser.add_argument('--list', nargs='*', help='List devices')
+    qvm.parser.add_argument(
+        '--attach',
+        nargs='*',
+        default=[],
+        help='List of devices to attach'
+    )
+    qvm.parser.add_argument(
+        '--detach',
+        nargs='*',
+        default=[],
+        help='List of devices to detach'
+    )
+
+    args = qvm.parse_args(vmname, *varargs, **kwargs)
+
+    current_devices = []
+    for device_type in ['pci', 'block', 'usb', 'bridge', 'mic']:
+        for device in args.vm.devices[device_type].assignments():
```

Comment:
Enumerating devices of unknown type (for example if relevant addon isn't installed in dom0) will fail. Please wrap it in try/except for each device type. Or better, implement https://github.com/QubesOS/qubes-issues/issues/5213 first and properly enumerate device types.

Reply:
I will implement it.